File: ts-go-runtypes/internal/compiler/resolver/check_unknowns_test.go
package resolver_test

import (
	"strings"
	"testing"

	"github.com/mionkit/ts-runtypes/internal/cachegen/operations"
	"github.com/mionkit/ts-runtypes/internal/protocol"
)

// End-to-end coverage for the `{checkUnknowns: true}` fused validators: the
// call site must route to the vst / vest FAMILIES (not a variant of val / verr),
// those families must render their own entries for NESTED NAMED types, and the
// emitted bodies must carry the key check placed after the property checks.
//
// The nested-named-type assertion is the load-bearing one. A compile-time
// variant keeps the plain family's InnerPrefix, so a named nested type would be
// dependency-called on the PLAIN entry and its extra keys would go unchecked —
// the exact regression this design avoids.

// familyPrefix derives an entry-key prefix from the operation registry rather
// than hardcoding a hash, so a future FnHashLen bump cannot turn a real
// regression into a confusing "no entry emitted" failure.
func familyPrefix(t *testing.T, opName string) string {
	t.Helper()
	op, ok := operations.ByName(opName)
	if !ok {
		t.Fatalf("%s operation missing from the registry", opName)
	}
	return operations.PlainHash(op.Name) + "_"
}

func countEntriesWithPrefix(modules map[string]string, prefix string) int {
	count := 0
	for _, mod := range modules {
		count += strings.Count(mod, prefix)
	}
	return count
}

func TestCheckUnknowns_RoutesToFusedFamily(t *testing.T) {
	modules := scanEntryModules(t, `import {createValidateFn} from '@ts-runtypes/core';
interface User {a: string; b: number}
export const isUser = createValidateFn<User>(undefined, {checkUnknowns: true});
`)
	if _, ok := findEntryWith(modules, familyPrefix(t, "validateStrict")); !ok {
		t.Fatalf("no validateStrict entry emitted\nmodules: %v", keys(modules))
	}
}

func TestCheckUnknowns_PlainCallStillUsesPlainFamily(t *testing.T) {
	modules := scanEntryModules(t, `import {createValidateFn} from '@ts-runtypes/core';
interface User {a: string; b: number}
export const isUser = createValidateFn<User>();
`)
	if _, ok := findEntryWith(modules, familyPrefix(t, "validate")); !ok {
		t.Fatalf("no plain validate entry emitted\nmodules: %v", keys(modules))
	}
	if _, ok := findEntryWith(modules, familyPrefix(t, "validateStrict")); ok {
		t.Fatalf("plain call site emitted a validateStrict entry — the flag leaked\nmodules: %v", keys(modules))
	}
}

// The design's whole point: a NAMED nested type is emitted as its own entry and
// dependency-called, so the fused family must render a fused entry for it too.
// Two vst entries (the root and the nested type) prove the recursion; a variant
// would produce one.
func TestCheckUnknowns_RendersFusedEntryForNestedNamedType(t *testing.T) {
	modules := scanEntryModules(t, `import {createValidateFn} from '@ts-runtypes/core';
interface Address {street: string; city: string}
interface Person {name: string; address: Address}
export const isPerson = createValidateFn<Person>(undefined, {checkUnknowns: true});
`)
	prefix := familyPrefix(t, "validateStrict")
	if got := countEntriesWithPrefix(modules, prefix); got < 2 {
		t.Fatalf("expected fused entries for BOTH Person and the nested Address, found %d occurrences of %q\nmodules: %v",
			got, prefix, keys(modules))
	}
}

// The key check must sit AFTER the property checks in the object's && chain.
// That ordering is what makes the O(1) key-count compare sound at every depth
// (every declared prop is known present by the time it runs), so it is a
// correctness property, not formatting.
func TestCheckUnknowns_AllRequiredShapeUsesKeyCountCompare(t *testing.T) {
	modules := scanEntryModules(t, `import {createValidateFn} from '@ts-runtypes/core';
interface User {a: string; b: number}
export const isUser = createValidateFn<User>(undefined, {checkUnknowns: true});
`)
	name, ok := findEntryWith(modules, familyPrefix(t, "validateStrict"))
	if !ok {
		t.Fatalf("no validateStrict entry emitted\nmodules: %v", keys(modules))
	}
	body := modules[name]
	if !strings.Contains(body, "=== 2") {
		t.Errorf("all-required 2-prop shape did not use the key-count compare:\n%s", body)
	}
	// The count compare comes last: the property checks it depends on precede it.
	countAt := strings.Index(body, "=== 2")
	propAt := strings.LastIndex(body, "typeof")
	if countAt >= 0 && propAt >= 0 && countAt < propAt {
		t.Errorf("key-count compare emitted BEFORE the property checks — the fast path is unsound there:\n%s", body)
	}
}

// Optional props make the count meaningless (a missing optional and an extra key
// both shift it), so those shapes must fall back to the key-list scan.
func TestCheckUnknowns_OptionalPropShapeUsesKeyScan(t *testing.T) {
	modules := scanEntryModules(t, `import {createValidateFn} from '@ts-runtypes/core';
interface User {a: string; b?: number}
export const isUser = createValidateFn<User>(undefined, {checkUnknowns: true});
`)
	name, ok := findEntryWith(modules, familyPrefix(t, "validateStrict"))
	if !ok {
		t.Fatalf("no validateStrict entry emitted\nmodules: %v", keys(modules))
	}
	body := modules[name]
	if strings.Contains(body, "countEnumKeys") {
		t.Errorf("optional-prop shape wrongly used the key-count fast path:\n%s", body)
	}
	if !strings.Contains(body, "hasUnknownKeysFromArray") {
		t.Errorf("optional-prop shape did not fall back to the key-list scan:\n%s", body)
	}
}

// An index signature declares every key matching it, so there is nothing to
// reject and the node takes no key check at all.
func TestCheckUnknowns_IndexSignatureTakesNoKeyCheck(t *testing.T) {
	modules := scanEntryModules(t, `import {createValidateFn} from '@ts-runtypes/core';
export const isRecord = createValidateFn<Record<string, number>>(undefined, {checkUnknowns: true});
`)
	name, ok := findEntryWith(modules, familyPrefix(t, "validateStrict"))
	if !ok {
		t.Fatalf("no validateStrict entry emitted\nmodules: %v", keys(modules))
	}
	body := modules[name]
	for _, unwanted := range []string{"countEnumKeys", "hasUnknownKeysFromArray"} {
		if strings.Contains(body, unwanted) {
			t.Errorf("index-signature shape emitted %q — every matching key IS declared:\n%s", unwanted, body)
		}
	}
}

func TestCheckUnknowns_ErrorsRoutesToFusedFamily(t *testing.T) {
	modules := scanEntryModules(t, `import {createGetValidationErrorsFn} from '@ts-runtypes/core';
interface User {a: string}
export const errorsOf = createGetValidationErrorsFn<User>(undefined, {checkUnknowns: true});
`)
	if _, ok := findEntryWith(modules, familyPrefix(t, "validationErrorsStrict")); !ok {
		t.Fatalf("no validationErrorsStrict entry emitted\nmodules: %v", keys(modules))
	}
}

// The fused option composes with the other compile-time options rather than
// replacing them: the axis is unchanged, so a variant of the FUSED family is
// what a combined call site must resolve to.
func TestCheckUnknowns_ComposesWithValidateOptions(t *testing.T) {
	modules := scanEntryModules(t, `import {createValidateFn} from '@ts-runtypes/core';
interface User {a: 'x'}
export const isUser = createValidateFn<User>(undefined, {checkUnknowns: true, noLiterals: true});
`)
	op, ok := operations.ByName("validateStrict")
	if !ok {
		t.Fatalf("validateStrict operation missing from the registry")
	}
	want := operations.FnHashFor(op, []string{"noLiterals"}, "", false) + "_"
	if _, found := findEntryWith(modules, want); !found {
		t.Fatalf("no noLiterals variant of the fused family emitted\nmodules: %v", keys(modules))
	}
}

// Marker test coverage rule (ts-go-runtypes/CLAUDE.md): both getRunTypeId call
// shapes, written as paired tests using the natural shape for each intent.

func TestCheckUnknowns_MarkerStaticForm(t *testing.T) {
	r := setupInline(t, map[string]string{"static.ts": `import {getRunTypeId} from '@ts-runtypes/core';
interface User {a: string; b: number}
export const id = getRunTypeId<User>();
`})
	resolved := resolveFile(t, r, "static.ts")
	if resolved.ID == "" {
		t.Fatalf("static getRunTypeId<User>() resolved no id")
	}
}

func TestCheckUnknowns_MarkerReflectForm(t *testing.T) {
	r := setupInline(t, map[string]string{"reflect.ts": `import {getRunTypeId} from '@ts-runtypes/core';
interface User {a: string; b: number}
const v: User = {a: 'x', b: 1};
export const id = getRunTypeId(v);
`})
	resolved := resolveFile(t, r, "reflect.ts")
	if resolved.ID == "" {
		t.Fatalf("reflect getRunTypeId(v) resolved no id")
	}
}

// The hash-equivalence assertion the rule requires: both shapes of the same
// type resolve to one entry, so a fused validator built from either reaches the
// same compiled function.
func TestCheckUnknowns_MarkerFormEquivalence(t *testing.T) {
	r := setupInline(t, map[string]string{
		"static.ts": `import {getRunTypeId} from '@ts-runtypes/core';
interface User {a: string; b: number}
export const id = getRunTypeId<User>();
`,
		"reflect.ts": `import {getRunTypeId} from '@ts-runtypes/core';
interface User {a: string; b: number}
const v: User = {a: 'x', b: 1};
export const id = getRunTypeId(v);
`,
	})
	static := resolveFile(t, r, "static.ts")
	reflect := resolveFile(t, r, "reflect.ts")
	if static.ID != reflect.ID {
		t.Fatalf("static vs reflect form of User disagree: %q vs %q", static.ID, reflect.ID)
	}
}

// A `checkUnknowns` site and a plain site for the SAME type must coexist: the
// two families are independent, so both entries ship (pay-for-use).
func TestCheckUnknowns_CoexistsWithPlainSiteForSameType(t *testing.T) {
	modules := scanEntryModules(t, `import {createValidateFn} from '@ts-runtypes/core';
interface User {a: string; b: number}
export const isUser = createValidateFn<User>();
export const isUserStrict = createValidateFn<User>(undefined, {checkUnknowns: true});
`)
	for _, opName := range []string{"validate", "validateStrict"} {
		if _, ok := findEntryWith(modules, familyPrefix(t, opName)); !ok {
			t.Fatalf("no %s entry emitted when both call shapes are present\nmodules: %v", opName, keys(modules))
		}
	}
}

// Nothing about the fused families may disturb the wire: a scan carrying only a
// plain site must not gain fused demand.
func TestCheckUnknowns_NoFusedDemandWithoutTheFlag(t *testing.T) {
	r := setupInline(t, map[string]string{"a.ts": `import {createValidateFn} from '@ts-runtypes/core';
interface User {a: string}
export const isUser = createValidateFn<User>();
`})
	resp := r.Dispatch(protocol.Request{Op: protocol.OpScanFiles, Files: []string{"a.ts"}, IncludeEntryModules: true})
	if resp.Error != "" {
		t.Fatalf("scan: %s", resp.Error)
	}
	for _, site := range resp.Sites {
		for _, demand := range site.Demand {
			if demand.FamilyTag == "vst" || demand.FamilyTag == "vest" {
				t.Fatalf("plain site demanded the fused family %q", demand.FamilyTag)
			}
		}
	}
}

// The fused validators run their key check INSIDE their own object guard, so
// they must not also pull in the one the standalone unknown-keys families carry.
//
// Both halves already avoid it — validate passes keepObjectCheck=false to
// callCheckUnknownPropertiesForHas, and the fused error arm calls
// emitParentUnknownKeyErrors directly rather than emitObjectUnknownKeyErrors,
// whose body IS guarded (that family has nothing above it asserting shape). But
// that is a property of where the shared helper's boundary happens to fall, not
// something the type system enforces: moving the guard down into the helper, or
// pointing the fused arm at the family arm, would silently emit the guard twice
// on every object node. Cheap to pin, easy to regress.
func TestCheckUnknowns_DoesNotDoubleGuardObjects(t *testing.T) {
	for _, row := range []struct{ label, factory, opName string }{
		{"validate", "createValidateFn", "validateStrict"},
		{"validationErrors", "createGetValidationErrorsFn", "validationErrorsStrict"},
	} {
		t.Run(row.label, func(t *testing.T) {
			modules := scanEntryModules(t, `import {`+row.factory+`} from '@ts-runtypes/core';
interface User {a: string; b?: number}
export const fn = `+row.factory+`<User>(undefined, {checkUnknowns: true});
`)
			name, ok := findEntryWith(modules, familyPrefix(t, row.opName))
			if !ok {
				t.Fatalf("no %s entry emitted\nmodules: %v", row.opName, keys(modules))
			}
			// The optional prop puts this shape on the key-SCAN path, which is the
			// one that could inherit a guard; the count-compare path emits none.
			body := modules[name]
			if got := strings.Count(body, "!==null&&!Array.isArray"); got > 1 {
				t.Errorf("%s emitted the object guard %d times, expected at most 1:\n%s", row.opName, got, body)
			}
			if got := strings.Count(body, "!== null && !Array.isArray"); got > 1 {
				t.Errorf("%s emitted the object guard %d times, expected at most 1:\n%s", row.opName, got, body)
			}
		})
	}
}

File: packages/ts-runtypes/test/features/checkUnknowns.test.ts
// End-to-end tests for the `{checkUnknowns: true}` fused validators — one
// compiled function that checks properties AND undeclared keys in a single walk,
// replacing `isT(v) && !hasUnknownKeys(v)`.
//
// The load-bearing test here is the PARITY suite: whatever shape is thrown at it,
// the fused validator must agree with the two-call composition it replaces. That
// is the contract users are trading their two calls for. The nested-named-type
// case is the one that would silently regress if the feature were built as a
// compile-time VARIANT instead of its own family, since a variant only reaches
// the root object.

import {describe, expect, it} from 'vitest';
import {createGetValidationErrorsFn, createHasUnknownKeysFn, createUnknownKeyErrorsFn, createValidateFn} from '@ts-runtypes/core';

describe('checkUnknowns — createValidateFn', () => {
  it('accepts a value with exactly the declared keys', () => {
    const isUser = createValidateFn<{a: string; b: number}>(undefined, {checkUnknowns: true});
    expect(isUser({a: 'x', b: 1})).toBe(true);
  });

  it('rejects an extra key the plain validator accepts', () => {
    const isUser = createValidateFn<{a: string; b: number}>(undefined, {checkUnknowns: true});
    const isUserLoose = createValidateFn<{a: string; b: number}>();
    expect(isUserLoose({a: 'x', b: 1, extra: true})).toBe(true);
    expect(isUser({a: 'x', b: 1, extra: true})).toBe(false);
  });

  it('still rejects a type mismatch', () => {
    const isUser = createValidateFn<{a: string; b: number}>(undefined, {checkUnknowns: true});
    expect(isUser({a: 'x', b: 'not a number'})).toBe(false);
  });

  // The whole reason this ships as a family rather than a variant: a NAMED
  // nested type is emitted as its own entry and dependency-called, so a
  // root-scoped variant would leave it checked by the plain validator and miss
  // the extra key entirely.
  it('rejects an extra key on a NAMED nested type', () => {
    type Address = {street: string; city: string};
    type Person = {name: string; address: Address};
    const isPerson = createValidateFn<Person>(undefined, {checkUnknowns: true});
    expect(isPerson({name: 'Ada', address: {street: 'A', city: 'B'}})).toBe(true);
    expect(isPerson({name: 'Ada', address: {street: 'A', city: 'B', zip: '1'}})).toBe(false);
  });

  it('rejects an extra key on an inline nested object', () => {
    const isPerson = createValidateFn<{name: string; address: {street: string}}>(undefined, {checkUnknowns: true});
    expect(isPerson({name: 'Ada', address: {street: 'A', extra: 1}})).toBe(false);
  });

  it('rejects an extra key inside an array element', () => {
    const isList = createValidateFn<{items: {id: number}[]}>(undefined, {checkUnknowns: true});
    expect(isList({items: [{id: 1}, {id: 2}]})).toBe(true);
    expect(isList({items: [{id: 1}, {id: 2, extra: true}]})).toBe(false);
  });

  it('accepts any key on an index-signature shape', () => {
    // Every key matching the index IS declared, so nothing can be "unknown".
    const isRecord = createValidateFn<Record<string, number>>(undefined, {checkUnknowns: true});
    expect(isRecord({a: 1, b: 2, anythingElse: 3})).toBe(true);
  });

  it('handles a shape with optional properties', () => {
    const isUser = createValidateFn<{a: string; b?: number}>(undefined, {checkUnknowns: true});
    expect(isUser({a: 'x'})).toBe(true);
    expect(isUser({a: 'x', b: 1})).toBe(true);
    expect(isUser({a: 'x', extra: true})).toBe(false);
  });

  it('leaves the plain validator untouched', () => {
    const isUserLoose = createValidateFn<{a: string}>();
    expect(isUserLoose({a: 'x', extra: 1})).toBe(true);
  });
});

describe('checkUnknowns — createGetValidationErrorsFn', () => {
  it('reports no errors for a clean value', () => {
    const errorsOf = createGetValidationErrorsFn<{a: string}>(undefined, {checkUnknowns: true});
    expect(errorsOf({a: 'x'})).toEqual([]);
  });

  it('reports an undeclared key as expected never', () => {
    const errorsOf = createGetValidationErrorsFn<{a: string}>(undefined, {checkUnknowns: true});
    expect(errorsOf({a: 'x', extra: 1})).toEqual([{path: ['extra'], expected: 'never'}]);
  });

  it('reports an undeclared key on a NAMED nested type with the right path', () => {
    type Address = {street: string};
    type Person = {name: string; address: Address};
    const errorsOf = createGetValidationErrorsFn<Person>(undefined, {checkUnknowns: true});
    expect(errorsOf({name: 'Ada', address: {street: 'A', zip: '1'}})).toEqual([{path: ['address', 'zip'], expected: 'never'}]);
  });

  it('reports both a type error and an undeclared key', () => {
    const errorsOf = createGetValidationErrorsFn<{a: string}>(undefined, {checkUnknowns: true});
    const errors = errorsOf({a: 1, extra: true});
    expect(errors).toContainEqual({path: ['a'], expected: 'string'});
    expect(errors).toContainEqual({path: ['extra'], expected: 'never'});
  });
});

// Error ORDER is a deliberate, documented divergence from the two-call form:
// `verr(v).concat(uke(v))` groups every type error ahead of every unknown-key
// error, but one walk cannot produce that grouping. The fused report interleaves
// per node, in walk order. Pinned here so a future change to the emit has to
// decide the order on purpose rather than drift into it.
describe('checkUnknowns — error order', () => {
  it('interleaves per node in walk order rather than grouping by kind', () => {
    type Inner = {x: string};
    type Outer = {inner: Inner; y: string};
    const errorsOf = createGetValidationErrorsFn<Outer>(undefined, {checkUnknowns: true});
    const errors = errorsOf({inner: {x: 1, innerExtra: true}, y: 2, outerExtra: true});

    // The inner node is fully reported (its type error AND its unknown key)
    // before the walk returns to the outer node's remaining property.
    expect(errors).toEqual([
      {path: ['inner', 'x'], expected: 'string'},
      {path: ['inner', 'innerExtra'], expected: 'never'},
      {path: ['y'], expected: 'string'},
      {path: ['outerExtra'], expected: 'never'},
    ]);
  });
});

// The contract users are trading their two calls for. Every case below is
// checked against the composition it replaces rather than against a hand-written
// expectation, so the two can never drift apart silently.
describe('checkUnknowns — parity with the two-call composition', () => {
  type Address = {street: string; city: string};
  type Person = {name: string; age: number; address: Address; tags?: string[]};

  const isPersonStrict = createValidateFn<Person>(undefined, {checkUnknowns: true});
  const isPerson = createValidateFn<Person>();
  const hasUnknown = createHasUnknownKeysFn<Person>();

  const errorsStrict = createGetValidationErrorsFn<Person>(undefined, {checkUnknowns: true});
  const typeErrors = createGetValidationErrorsFn<Person>();
  const keyErrors = createUnknownKeyErrorsFn<Person>();

  // Values the object guard ADMITS (`typeof v === 'object' && v !== null`).
  // Arrays, Maps and Dates belong here, not with the primitives: they pass that
  // gate for a required-prop shape, so their own enumerable keys really are
  // undeclared ones. `[{name: 'Ada'}]` carrying a `'0'` key is a genuine
  // `{expected: 'never'}` entry, not a false positive.
  const objectCorpus: unknown[] = [
    {name: 'Ada', age: 36, address: {street: 'A', city: 'B'}},
    {name: 'Ada', age: 36, address: {street: 'A', city: 'B'}, tags: ['x']},
    {name: 'Ada', age: 36, address: {street: 'A', city: 'B'}, extra: true},
    {name: 'Ada', age: 36, address: {street: 'A', city: 'B', zip: '1'}},
    {name: 'Ada', age: 'not a number', address: {street: 'A', city: 'B'}},
    {name: 'Ada', age: 36, address: {street: 'A'}},
    {name: 'Ada', age: 36, address: {street: 'A', city: 'B', zip: '1'}, extra: 1},
    {},
    [],
    [{name: 'Ada'}],
    new Map(),
    new Date(),
  ];

  // Values the object guard REJECTS. Since the unknown-keys families gained
  // their own shape guard (they now answer [] / false for a value the schema
  // does not admit, rather than throwing or inventing one entry per character
  // index), the ERROR composition is well-defined over these too — so they ride
  // both oracles below rather than only the boolean one.
  const primitiveCorpus: unknown[] = [null, undefined, 'a string', 42];

  it.each([...objectCorpus, ...primitiveCorpus].map((value, index) => [index, value] as const))(
    'validator agrees with isT(v) && !hasUnknown(v) — case %i',
    (_index, value) => {
      expect(isPersonStrict(value)).toBe(isPerson(value) && !hasUnknown(value));
    }
  );

  // Compared as SETS: the fused walk interleaves entries where the two-call form
  // groups them (see the error-order suite above), so membership is the shared
  // contract, not sequence.
  //
  // This oracle used to be restricted to guard-admitted values, because
  // `createUnknownKeyErrorsFn` descended into declared properties with nothing
  // asserting shape — it threw on `null` and returned one bogus
  // `{expected: 'never'}` per character index on a string. That is fixed, so the
  // reference is defined over every input and the corpus is whole again.
  it.each([...objectCorpus, ...primitiveCorpus].map((value, index) => [index, value] as const))(
    'error report matches verr + uke as a set — case %i',
    (_index, value) => {
      const fused = errorsStrict(value);
      const composed = [...typeErrors(value), ...keyErrors(value)];
      const sortKey = (error: {path: unknown[]; expected: string}) => JSON.stringify([error.path, error.expected]);
      expect(fused.map(sortKey).sort()).toEqual(composed.map(sortKey).sort());
    }
  );

  // A value the object guard rejects never reaches the key check, so the fused
  // report must be exactly the plain type-error report, and must never throw.
  it.each(primitiveCorpus.map((value, index) => [index, value] as const))(
    'guard-rejected input reports only the type error — case %i',
    (_index, value) => {
      expect(() => errorsStrict(value)).not.toThrow();
      expect(errorsStrict(value)).toEqual(typeErrors(value));
    }
  );
});
